Add Magento 2 docker-compose

supercid opened the pull request:
- [ ] Write docs
- [ ] Write a build script to push to docker registry
- [ ] Mount platforms volumes locally for better IDE completion
- [ ] Add HTTPS with let's encrypt in traefik. I started this but couldn't configure let's encrypt properly
- [ ] Use the same network as Playcart, so all services can run under port:80
- [x] Use traefik as a standalone docker-compose file, this will allow us to add mode images for other platforms later on. Then you can build/tear down as you wish/on demand
- [ ] Add a small graph about image dependencies. They should all be open sourced
- [ ] Move xdebug on/off script to php base image. Enable whenever needed.
- [ ] Add version as arg when building images for easier tagging/versioning
- [ ] Remove `@TODOS` from codebase


### Docs to be written
- [ ] How to push to docker registry
- [ ] Commands you can run with traefik for configuration
- [ ] How to run Magento commands without logging into the container
- [ ] How to configure PHPStorm to listen to XDebug (Needs to be listening before starting the container => those are pretty good instructions: https://gist.github.com/manuelselbach/8a214ae012964b1d49d9fb019f5f5d7b
- [ ] ??

supercid committed:
Add initial Magento 2 docker-compose files

supercid committed:
Move PHP to a standalone image

supercid committed:
Add Prestashop image

supercid committed:
Add PHP 7.3 image

supercid committed:
Add traefik docker-compose file

supercid committed:
Remove networks from prestashop docker-compose file and use traefik external one

supercid committed:
Remove traefik config from magento 2 image

- Fix routing
- Use same external network as traefik

supercid committed:
Upgrade Magento base image to use 2.4.0

supercid committed:
Add instructions to readme

supercid committed:
Magento2 - Disable 2FA by default

mridang commented on magento2/magento2/scripts/init-and-run.sh:
```diff
@@ -0,0 +1,110 @@
+if [ ! -f /var/www/html/magento2/.installed ]; then
+  cd /var/www/html/magento2/ || exit
+  composer config --global repos.packagist composer https://packagist.org
+  composer require nosto/php-sdk:@stable
```
Why is the SDK set up here?

supercid replied:
Magento needs to have the SDK before requiring the main extension

mridang commented on README.md:
````diff
@@ -1 +1,44 @@
+Make sure you have created the default external network
+```bash
+	cd traefik
+	docker-compose up -d
````
Assuming that I don't want to run Traefik - what do I need to do to run Magento 2?

supercid replied:
- Create the networks as usual
- Expose port 80 in the Magento docker-compose file

I'll write about this

mridang replied:
I would then need to shut my Play wouldn't I? I cannot do that. Can we change the ports?

supercid replied:
For the time being, yes. If you'd like to change the ports, do so on traefik, exposing 80 to some other free port you have.
You need to update the platform hostname as well.

mridang replied:
I'm a bit lost. Any examples?

supercid replied:
Updated to use 8081 by default

mridang commented on magento2/docker-compose.yml:
```diff
@@ -0,0 +1,99 @@
+  ####          MariaDB Container          ######
+  ##############################################
+  db_m2:
+    image: nosto/mariadb:10.0
```
Why do we use a Nosto one and not https://hub.docker.com/layers/mariadb/library/mariadb/10.5.5/images/sha256-a7803e8939d15e80ac32eaea4b072884b3d335133fb679c7cd103fad75dafa9c?context=explore

supercid replied:
`I have increased inno_db values to get rid of some warning when the indexer
runs with default configuration`

mridang commented on magento2/docker-compose.yml:
```diff
@@ -0,0 +1,99 @@
+  ####          RabbitMQ Container         ######
+  ##############################################
+  rabbitmq_m2:
+    image: rabbitmq:3.7-management
```
Why is this not https://hub.docker.com/layers/rabbitmq/library/rabbitmq/3.7/images/sha256-2fb8f0da539258741eae28031e533070660033c7b8698051c420be41b47b944b?context=explore

mridang replied:
Do we need the management tools?

supercid replied:
Easier to check how the queues are running/easier debugging

mridang replied:
👍

supercid commented:
I have increased inno_db values to get rid of some warning when the indexer
runs with default configuration

On Tue 18. Aug 2020 at 12.12, Mridang Agarwalla <notifications@github.com>
wrote:

>
>
> *@mridang* commented on this pull request.
>
>
>
>
> ------------------------------
>
>
>
>
> In magento2/docker-compose.yml
> <https://github.com/Nosto/docker-platforms/pull/1#discussion_r472034446>:
>
>
> > @@ -0,0 +1,99 @@
>
> +version: '3.3'
>
> +
>
> +services:
>
> +  ################################################
>
> +  ####          MariaDB Container          ######
>
> +  ##############################################
>
> +  db_m2:
>
> +    image: nosto/mariadb:10.0
>
>
>
> Why do we use a Nosto one and not
> https://hub.docker.com/layers/mariadb/library/mariadb/10.5.5/images/sha256-a7803e8939d15e80ac32eaea4b072884b3d335133fb679c7cd103fad75dafa9c?context=explore
>
>
>
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Nosto/docker-platforms/pull/1#pullrequestreview-469139423>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVGNRCYFVHGOKBYUXHPBXLSBJAYNANCNFSM4PGTZTWA>
> .
>
>
> --
Cid Lopes
Mobile: +358 40 706 4443

mridang commented on magento2/magento-base/Dockerfile:
```diff
@@ -1,45 +1,12 @@
-FROM php:7.2-apache
+FROM nosto/php-apache:7.3
```
Why the Nosto ones? Just curious.

supercid replied:
Magento uses multiple libraries and extensions that are trimmed from the original images. Such as:
sockets, intl, opcache, mbstring etc.
Those are installed in this custom image

mridang commented on magento2/magento2/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+RUN echo 'www-data ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
+
+# Install XDebug && APCu
+RUN pecl install xdebug apcu && \
```
Why isn't this in the base image?

supercid replied:
Moved there

mridang commented on magento2/magento2/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+ADD scripts/init-and-run.sh /usr/local/bin/init-and-run
+ADD scripts/xdebug.sh /usr/local/bin/xdebug
+
+RUN echo 'www-data ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
```
Why isn't this in the base image?

supercid replied:
Moved there

mridang commented on magento2/magento2/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+    chmod +x /usr/local/bin/gosu
+#@TODO: Might not be needed and can be removed to slim the image
+
+USER www-data
```
Why isn't this in the base image?

supercid replied:
We want to run commands as root while building the image. Leave apache user for when you use the image, so you can run Magento/any platform command without having to specify permissions afterwards.

mridang commented on magento2/.docker/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+
+USER www-data
+
+RUN composer global require hirak/prestissimo
```
Shouldn't this be in the base PHP image?

supercid replied:
Moved there

mridang commented on magento2/.docker/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+RUN curl -o /usr/local/bin/gosu -sSL "https://github.com/tianon/gosu/releases/download/1.6/gosu-$(dpkg --print-architecture)" && \
+    chmod +x /usr/local/bin/gosu
+
+USER www-data
```
Shouldn't this be in the base Apache image?

supercid replied:
File removed

mridang commented on magento2/.docker/generate-code.sh:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash -x
+
+# Run this from the root dir of the repo with ./.docker/generate-code.sh
+docker run -d -it --name mage_code_gen \
```
If the container is already running, why does this start a new one?

supercid replied:
File removed, this is unnecessary as will be mounted as a volume

mridang commented on magento2/magento2/scripts/init-and-run.sh:
```diff
@@ -0,0 +1,110 @@
+    sleep 5
+  done
+
+  until curl -i ${ES_HOST}:9200/_cluster/health; do
```
This should be done in the docker-compose using a health-check https://stackoverflow.com/questions/31746182/docker-compose-wait-for-container-x-before-starting-y

mridang commented on magento2/magento2/scripts/init-and-run.sh:
```diff
@@ -0,0 +1,110 @@
+  composer config --global repos.packagist composer https://packagist.org
+  composer require nosto/php-sdk:@stable
+
+  until mysql -u"${MYSQL_USER}" -p"${MYSQL_PASSWORD}" -h"${MYSQL_HOST}"; do
```
This should be done in the docker-compose using a health-check https://stackoverflow.com/questions/31746182/docker-compose-wait-for-container-x-before-starting-y

supercid replied:
I couldn't configure that using docker-compose file version 3.
In order to move forward with this PR, there's an open issue here for later improvement:
https://github.com/Nosto/docker-platforms/issues/2

mridang commented on prestashop/scripts/init-and-run.sh:
```diff
@@ -0,0 +1,47 @@
+
+if [ ! -f /var/www/html/prestashop/.installed ]; then
+  
+  until mysql -u"${MYSQL_USER}" -p"${MYSQL_PASSWORD}" -h"${MYSQL_HOST}"; do
```
This should be done in the docker-compose using a health-check https://stackoverflow.com/questions/31746182/docker-compose-wait-for-container-x-before-starting-y

mridang commented on prestashop/prestashop-base/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+
+ENV PRESTASHOP_VERSION 1.7.6.7
+ENV PRESTASHOP_ARCHIVE prestashop_${PRESTASHOP_VERSION}
+ENV TERM xterm
```
Shouldn't this be in the base image?

supercid replied:
Removed

mridang commented on magento2/.docker/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+ADD scripts/init-and-run.sh /usr/local/bin/init-and-run
+ADD scripts/xdebug.sh /usr/local/bin/xdebug
+
+RUN echo 'www-data ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
```
Shouldn't this be in the Apache image or the base image?

supercid replied:
Moved there

mridang commented on magento2/.docker/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+RUN echo 'www-data ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
+
+# Install XDebug && APCu
+RUN pecl install xdebug apcu && \
```
Shouldn't this be in base PHP image?

mridang commented on php-7.2-apache/Dockerfile:
```diff
@@ -0,0 +1,37 @@
+ENV COMPOSER_ALLOW_SUPERUSER 1
+
+RUN apt-get update && \
+    apt-get -y -qq install apt-transport-https locales jq unzip wget \
```
Can you reformat to be in separate lines? One dependency per line. I normally just run this multiple times and then flatten the layers.

`RUN apt-get -y -qq install <dep>`

It is more readable but I'm not sure what are the other drawbacks.

mridang commented on php-7.3-apache/Dockerfile:
```diff
@@ -0,0 +1,37 @@
+ENV TERM=xterm
+ENV COMPOSER_ALLOW_SUPERUSER 1
+
+RUN apt-get update && \
```
Can you reformat to be in separate lines? One dependency per line. I normally just run this multiple times and then flatten the layers.

`RUN apt-get -y -qq install <dep>`

It is more readable but I'm not sure what are the other drawbacks.

supercid committed:
Move traefik to run on port 8081

supercid committed:
Update README ports

supercid committed:
Move xdebug script to PHP images

Move PHP images under `php` dir

supercid committed:
Update Mariadb to 10.2

supercid committed:
Remove useless comments from prestashop docker-compose

supercid committed:
Magento 2 - Upgrade MariaDB

Remove xdebug from Dockerfile
Use volume for MySQL
Check for data before starting container to reinstall Magento

supercid committed:
Add Shopware Images

supercid committed:
Remove redundant dockerfile and script to build magento generated files

supercid committed:
Remove xterm env from Prestashop image

supercid committed:
Dump Magento installation to host before starting up containers

This is unpractical since loading times for mounted volumes on docker for mac are absurdly long.
